src/toast/pipeline_tools/binning.py

```diff
@@ -0,0 +1,258 @@
+
+    npix = 12 * args.nside * args.nside
+
+    invnpp = DistPixels(
+        comm=comm.comm_world,
+        size=npix,
+        nnz=6,
+        dtype=np.float64,
+        submap=subnpix,
+        local=localsm,
+    )
+    hits = DistPixels(
+        comm=comm.comm_world,
+        size=npix,
+        nnz=1,
+        dtype=np.int64,
+        submap=subnpix,
+        local=localsm,
+    )
+    zmap = DistPixels(
+        comm=comm.comm_world,
+        size=npix,
+        nnz=3,
+        dtype=np.float64,
+        submap=subnpix,
+        local=localsm,
+    )
+
+    # compute the hits and covariance once, since the pointing and noise
+    # weights are fixed.
+
+    if invnpp.data is not None:
+        invnpp.data.fill(0.0)
+
+    if hits.data is not None:
+        hits.data.fill(0)
+
+    build_invnpp = OpAccumDiag(detweights=detweights, invnpp=invnpp, hits=hits)
+    build_invnpp.exec(data)
+
+    if comm.world_rank == 0 and verbose:
+        timer.report_clear("Accumulate N_pp'^1")
+
+    invnpp.allreduce()
+    hits.allreduce()
+
+    if comm.world_rank == 0 and verbose:
+        timer.report_clear("All reduce N_pp'^1")
+
+    if args.write_hits:
+        fname = os.path.join(args.outdir, "hits.fits")
+        if args.zip_maps:
+            fname += ".gz"
+        hits.write_healpix_fits(fname)
+        if comm.world_rank == 0 and verbose:
+            log.info("Wrote hits to {}".format(fname))
+    if args.write_wcov_inv:
+        fname = os.path.join(args.outdir, "invnpp.fits")
+        if args.zip_maps:
+            fname += ".gz"
+        invnpp.write_healpix_fits(fname)
+        if comm.world_rank == 0 and verbose:
+            log.info("Wrote inverse white noise covariance to {}".format(fname))
+
+    if comm.world_rank == 0 and verbose:
+        timer.report_clear("Writing hits and N_pp'^1")
+
+    # invert it
+    covariance_invert(invnpp, 1.0e-3)
+
+    if comm.world_rank == 0 and verbose:
+        timer.report_clear("Invert N_pp'^1")
+
+    if args.write_wcov:
+        fname = os.path.join(args.outdir, "npp.fits")
+        invnpp.write_healpix_fits(fname)
+        if comm.world_rank == 0 and verbose:
+            log.info("Wrote white noise covariance to {}".format(fname))
+
+    if comm.world_rank == 0 and verbose:
+        timer.report_clear("Write N_pp'")
+
+    # in debug mode, print out data distribution information
+    if args.debug:
+        handle = None
+        if rank == 0:
+            handle = open(os.path.join(args.outdir, "distdata.txt"), "w")
+        data.info(handle)
+        if rank == 0:
+            handle.close()
+        if comm.world_rank == 0 and verbose:
+            timer.report_clear("Dumping data distribution")
+    return invnpp, zmap
+
+
+def apply_binner(
+    args,
+    comm,
+    data,
+    invnpp,
+    zmap,
+    detweights,
+    outpath,
+    cache_prefix=None,
+    prefix="binned",
+    verbose=True,
+):
+    """ Bin the signal in `cache_prefix` onto `zmap`
```

Comment:
what is a `zmap`?

Reply:
The distributed map. I revised the docstrings.